File: C++/Hard/Accepted/count-unique-characters-of-all-substrings-of-a-given-string_0.cpp
class Solution {
public:
    int uniqueLetterString(string s) {
        
        int n = s.size();
        vector<vector<int>> pos(26);
        for (int i = 0; i < n; ++i) {
            pos[s[i] - 'A'].push_back(i);
        }
        int ans = 0, mod = 1e9 + 7;
        for (int i = 0; i < 26; ++i) {
            for (int j = 0; j < pos[i].size(); ++j) {
                int prev = j > 0 ? pos[i][j - 1] : -1;
                int next = j < pos[i].size() - 1 ? pos[i][j + 1] : n;
                ans = (ans + (pos[i][j] - prev) * (next - pos[i][j])) % mod;
            }
        }
        return ans;
    }
};

File: Java/Hard/Accepted/count-unique-characters-of-all-substrings-of-a-given-string_0.java
class Solution {
    public int uniqueLetterString(String s) {
        
        int n = s.length();
        int[][] index = new int[26][2];
        for (int i = 0; i < 26; ++i) {
            Arrays.fill(index[i], -1);
        }
        int ans = 0;
        for (int i = 0; i < n; ++i) {
            int x = s.charAt(i) - 'A';
            ans += (i - index[x][1]) * (index[x][1] - index[x][0]);
            index[x] = new int[]{index[x][1], i};
        }
        for (int i = 0; i < 26; ++i) {
            ans += (n - index[i][1]) * (index[i][1] - index[i][0]);
        }
        return ans;
    }
}